compiler/lib/deadcode.ml

```diff
@@ -349,28 +349,29 @@ let remove_empty_blocks st (p : Code.program) : Code.program =
     if not (Code.cont_equal cont cont') then st.block_shortcut <- st.block_shortcut + 1;
     cont'
   in
-  Addr.Map.iter
-    (fun pc block ->
-      match block with
-      | { params; body; branch = Branch cont; _ } when empty_body body ->
-          let args =
-            List.fold_left
-              ~f:(fun args x -> Var.Set.add x args)
-              ~init:Var.Set.empty
-              (snd cont)
-          in
-          (* We can skip an empty block if its parameters are only
+  let register_block_if_empty pc block =
+    match block with
+    | { params; body; branch = Branch cont; _ } when empty_body body ->
+        let args =
+          List.fold_left
+            ~f:(fun args x -> Var.Set.add x args)
+            ~init:Var.Set.empty
+            (snd cont)
+        in
+        (* We can skip an empty block if its parameters are only
              used as argument to the continuation *)
-          if
-            List.for_all
-              ~f:(fun x -> st.live.(Var.idx x) = 1 && Var.Set.mem x args)
-              params
-          then Addr.Hashtbl.add shortcuts pc (params, cont)
-      | _ -> ())
-    p.blocks;
+        if List.for_all ~f:(fun x -> st.live.(Var.idx x) = 1 && Var.Set.mem x args) params
+        then Addr.Hashtbl.add shortcuts pc (params, cont)
+    | _ -> ()
+  in
+  Addr.Map.iter register_block_if_empty p.blocks;
+  let addr_map_rev_fold f m acc =
+    let l = Addr.Map.fold (fun pc block rem -> (pc, block) :: rem) m [] in
+    List.fold_left ~f l ~init:acc
+  in
   let blocks =
-    Addr.Map.fold
-      (fun pc block blocks ->
+    addr_map_rev_fold
```

Comment:
Are you relying on some correlation between Addr.t ordering and control flow ordering ? It seems fragile but maybe it's fine.is it too costly to use the proper ordering ? Can we at least have a comment about this ?

Reply:
Indeed, I'm relying on the fact that forward branches target blocks with higher addresses in the code generated by the OCaml compiler. This seems unlikely to change.
 I could do a depth-first traversal of the code, but this would be certainly more costly.